### ISSUE TITLE
Allow Skeleton3D bones to have shear transformations.

### ISSUE THREAD

**Issue by Goy288** [godotengine/godot-proposals#11176]
### Describe the project you are working on

I'm working on an addon that reconstructs 2D skeletal Polygon2D systems into 3D mesh equivalents with Skeleton3D.

### Describe the problem or limitation you are having in your project

In 2D skeletal systems, you are able to apply a skew value to a bone. Making it and all the connected Polygon2D vertices and bones properly skew.

Skeleton3D based systems cannot do this. Apply any value to a Bone's basis all you want, but at most it will rotate or scale a bit, never properly shearing. Applying the same basis modifications to a Node3D, and it'll work, but to a Skeleton3D bone? Nope.

Comparison:
![image](https://github.com/user-attachments/assets/f5742dba-f3c9-4b3c-98ff-9da213be31dc)
It's just a little taller than usual. Meanwhille...

![image](https://github.com/user-attachments/assets/f7dbdae9-8017-4723-bf95-d2af7ebbe066)
...The 2D version is applied exactly how you'd expect.

I don't know if this is technically a bug, or it was made this way for some reason that I cannot determine, but it's been a pain in the butt trying to find workarounds to this. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow Skeleton3D bones to properly interpret a basis's shear and apply them exactly how you'd expect them to, like in the above.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I mean, I'm not gonna pretend that I know how it all works under the hood, but I think something like this:
```gdscript
var local_3d_transform := skeleton3d.get_bone_pose(bone_idx)
var sheared_basis := Basis()

sheared_basis.x = Vector3(1, 0, 0)
sheared_basis.y = Vector3(skew_factor, 1, 0)
sheared_basis.z = Vector3(0, 0, 1)

local_3d_transform = Transform3D(sheared_basis, local_3d_transform.origin)

skeleton3d.set_bone_pose(bone_idx, local_3d_transform)
```
Should properly shear a bone, its children, and its weighted meshes.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Simplest way is this nonsensical workaround that I'm currently using that _doesn't look quite right_.

![image](https://github.com/user-attachments/assets/5259e76e-9d17-4237-a12c-a179ceaf28a3)
```gdscript
# Pseudo-skew
if skew_2d != 0:
	var skew_matrix := Basis()
	var skew_atan := atan(-skew_2d)
	skew_matrix.x.y = -skew_2d
	
	var skew_inv_lerp_y: float
	if skew_atan > 0:
		skew_inv_lerp_y = inverse_lerp(1,0,skew_atan)
	elif skew_atan < 0:
		skew_inv_lerp_y = inverse_lerp(-1,0,skew_atan)
	
	scale_2d.y *= skew_inv_lerp_y
	origin_2d.y *= skew_inv_lerp_y
	
	combined_basis = combined_basis * skew_matrix
```

...but obviously, this is not ideal. The only _other_ way that I think I could make this work, is to set up the bones to auto adjust themselves (and their children) to accurately mimic how they'd be positioned, scaled, and rotated if they could properly take in shear transformations. _Then_, I'd need to program a system that sends data to a vertex shader that accepts an array with skew values and any other needed transform values in an array, and _then_ have the vertex shader properly adjust the vertices to accurately resemble how it'd look in the Skeleton2D version.

Anyway, uh, more than "a few lines of script".

### Is there a reason why this should be core and not an add-on in the asset library?

Well, I think shearing should simply just properly _work_ when you apply it to a Skeleton3D bone. I'm like, half sure this is a bug, but I'm posting this issue in here to be safe in case if it's part of some design/optimization or whatever. Slight possibility that this is also a bug on my end, but I've tried every method I could to adjust the Transform3D's basis, so I'm like 99% sure it's the Skeleton3D's fault.

If nothing else, I could argue that this would allow more proper consistency between how Skeleton2D and Skeleton3D systems work, as one system currently can shear bones, but the other cannot.

**Comment by Goy288** [comment]
Well, yeah, but there's still applications for it, like wanting to add a wobble like effect via scripting without shaders, or making certain things more identifiable from a top-down angle. I can't say that there's a _gajillion_ different uses for it, but I don't think there's anything particularly wrong with implementing support for something niche like this, especially if it's technically a result of the current implementation being innacurrate.

On another note, been looking into the [source](https://github.com/godotengine/godot/blob/master/scene/3d/skeleton_3d.cpp#L845), and it seems that the reason why it ignores the shear transforms in the first place, is that it applies the pose data like this:
```cpp
bones[p_bone].pose_position = p_pose.origin;
bones[p_bone].pose_rotation = p_pose.basis.get_rotation_quaternion();
bones[p_bone].pose_scale = p_pose.basis.get_scale();
```
With the [header file](https://github.com/godotengine/godot/blob/master/scene/3d/skeleton_3d.h#L114) having this:
```cpp
void update_pose_cache() {
	if (pose_cache_dirty) {
		pose_cache.basis.set_quaternion_scale(pose_rotation, pose_scale);
		pose_cache.origin = pose_position;
		pose_cache_dirty = false;
	}
}
```
Makes sense why shearing doesn't work now, as it doesn't seem to really account for it at all. I could envision a way to fix it, but it'd definitely involve a bit of restructuring.

Hm, now that I think about it, I _could_ potentially implement a workaround for my rather specific skewing needs that wouldn't require _as much_ restructuring. Would require making a GDExtension (something that I'm currently not very familiar with), but it'd probably be easier than implementing weird vertex shader workarounds.

**Comment by Goy288** [comment]
Honestly, think I might hold off on even implementing the workaround in the first release.

The more I look at it, the more it seems that I'll likely need to make my custom node's generated Skeleton3D's bones non-hierarchal, and implement my own arbitrary hierarchy system. The reason being is that I want to implement bone transform overrides as an important feature, and I can't have that if I just have the Skeleton3D bones mimic the Bone2Ds individually without them being parented properly. I wanna focus on the first release for now (and maybe pray that the Skeleton3D system eventually gets updated so that I won't have to worry about fixing it later). I'll just add a disclaimer in the release that clarifies the Skeleton3D thing and why skewing is not yet currently implemented.

**Comment by TokageItLab** [comment]
This is one feature that was removed from 3.x to 4.0.

First, we need to consider the following two compatibility issues:
- glTF format does not allow shear
- Position/Rotation/Scale tracks have compress/optimize functionality, but they cannot handle shear. If streaming animation is implemented in the future, shear will be removed anyway since it must use compression

So shear should never be allowed, at least for Rest transforms and Pose transforms.

An idea that might be possible here is something like the custom pose that existed in Godot 3.x, with parameters for shear separate from rest and pose. However, it would not be easy to break it into the internal pose calculation.

Given these considerations, I think the only acceptable way to do this would be to have each bone have meta data for shear and pass it to the Shader parameter to be transformed through VertexShader. https://github.com/godotengine/godot-proposals/issues/9735 would help with that.

**Comment by Goy288** [comment]
Yeah, I've been thinking of manually passing transformation data as uniforms myself, though something like SHEAR_TRANSFORM built-ins would make the process substantially easier.

The remaining caveat I think is that shearing would also affect the position of _other_ child bone transforms. In my specific setup, I am fortunate enough to be basing this system off of just mimicking the position of various Bone2Ds, so I could easily get around this by having each 3D bone follow the 2D one's global position and place them non-hierarchly, instead of them just being properly parented to each other and only modifying their local position when needed. This would be less optimized than my original system, but it certainly *could* work I think.

But then, to be compatible with my node's built-in bone modifier functionality, I'd also need to implement my own kind of hacky bone hierarchy system so that the children of each bone are moved along with them. I think it's certainly doable, but it sure sounds like a headache lol
